samza-api/src/main/java/org/apache/samza/operators/functions/StreamExpander.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.functions;
+
+import java.io.Serializable;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.StreamGraph;
+import org.apache.samza.operators.descriptors.base.stream.InputDescriptor;
+
+/**
+ * Expands the provided {@link InputDescriptor} to a sub-DAG of one or more operators on the {@link StreamGraph},
+ * and returns a new {@link MessageStream} with the combined results. Called when {@link StreamGraph#getInputStream}
+ * is being used to get a {@link MessageStream} using an {@link InputDescriptor} from an expanding system descriptor.
+ *
+ * <p>
+ * This is provided by default by expanding system descriptor implementations and can not be overridden
+ * or set on a per stream level.
+ *
+ * @param <OM> type of the messages in the resultant {@link MessageStream}
+ */
+public interface StreamExpander<OM> extends Serializable {
```

Comment:
Yeah, discussed this with Cameron earlier. Adding lifecycle methods to these wouldn't be sufficient, since the function is instantiated in SystemDescriptor implementations which don't (shouldn't) have lifecycle methods. If we need to support lifecycle for expand functions, we'll change the planner to do a 2-phase graph construction: first build the user specified graph, then apply the expand functions. Until then, expanding system implementers can create and destroy context within the StreamExpander#apply call.